docs/posts/duckdb-for-rag/index.qmd

```diff
@@ -0,0 +1,512 @@
+performance (accuracy or efficiency) than first retrieving the relevant subsets
+of the textbook.
+:::
+
+Traditional RAG embeds text into vectors of numbers and then computes a
+similarity metric between input text and the knowledge base to determine the
+most relevant context.
+
+```{mermaid}
+graph LR
+    style T fill:#1E90FF,stroke:#1E90FF,color:#ffffff
+    style N fill:#1E90FF,stroke:#1E90FF,color:#ffffff
+    style C fill:#1E90FF,stroke:#1E90FF,color:#ffffff
+
+    T("text") --> N("numbers (embedding vector)")
+    N --> C("compute similarity metric")
+```
+
+Let's jump into the code!
+
+## RAG over Ibis documentation
+
+We'll use the Ibis documentation as our knowledge base.
+
+### Setup code
+
+First we need to install and import the necessary packages:
+
+```bash
+pip install 'ibis-framework[duckdb]' openai python-dotenv
+```
+
+```{python}
+import os  # <1>
+import ibis
+import shutil
+import subprocess
+
+from openai import OpenAI
+from dotenv import load_dotenv  # <1>
+```
+
+1. Import the libraries we'll use.
+
+We need to set the `OPENAI_API_KEY` environment variable to our OpenAI API key.
+Do this by creating a `.env` file in the same directory as the code and adding
+the following line:
+
+```txt
+OPENAI_API_KEY=<your-api-key>  # <1>
+```
+
+1. Replace with your OpenAI API key.
+
+Then load the environment variables:
+
+```{python}
+load_dotenv()  # <1>
+```
+
+1. Load environment variables from the `.env` file.
+
+Now we're ready to make OpenAI calls.
+
+Let's also setup Ibis for interactive use:
+
+```{python}
+ibis.options.interactive = True  # <1>
+```
+
+1. Turn on Ibis interactive mode.
+
+### Getting the data into DuckDB
+
+We'll use a function to clone a given repository:
+
+```{python}
+def clone_repo(repo_url):  # <1>
+    subprocess.run(["git", "clone", "--depth", "1", repo_url], check=False)  # <1>
+```
+
+1. Define a function to clone a repository.
+
+Let's clone `ibis` and prepare to ingest all the Quarto markdown files into a
+DuckDB database called `rag.ddb`:
+
+```{python}
+org_url = "https://github.com/ibis-project"  # <1>
+repo_name = "ibis"  # <1>
+
+shutil.rmtree(repo_name, ignore_errors=True)  # <2>
+clone_repo(f"{org_url}/{repo_name}")  # <3>
+
+if os.path.exists("rag.ddb"):  # <4>
+    os.remove("rag.ddb")
+if os.path.exists("rag.ddb.wal"):
+    os.remove("rag.ddb.wal")  # <4>
+
+rag_con = ibis.connect("duckdb://rag.ddb")  # <5>
+rag_con.list_tables()  # <6>
+```
+
+1. Setup the repository to clone.
+2. Remove the repository if it exists to ensure a fresh clone.
+3. Clone the repository.
+4. Remove the DuckDB database if it exists to ensure a fresh database.
+5. Connect to the DuckDB database.
+6. List the tables in the database, showing it is empty.
+
+We'll walk through all the cloned files. For `.qmd` files, read the contents and
+insert them into the `ibis` table:
+
+```{python}
+for root, _, files in os.walk(repo_name):  # <1>
```

Comment:
I think `Path(repo_name).rglob("*.qmd")` might be a less cluttered approach to recursively reading all `.qmd` files.